src/Illuminate/Console/Process/PendingProcess.php

```diff
@@ -0,0 +1,239 @@
+     *
+     * @var bool
+     */
+    protected $delayStart = false;
+
+    /**
+     * The factory instance.
+     *
+     * @var \Illuminate\Console\Process\Factory
+     */
+    protected $factory;
+
+    /**
+     * The process's output callback.
+     *
+     * @var (callable(string, int): mixed)|null
+     */
+    protected $output;
+
+    /**
+     * The process's path.
+     *
+     * @var string|null
+     */
+    protected $path;
+
+    /**
+     * The process's timeout.
+     *
+     * @var float|null
+     */
+    protected $timeout = 60.0;
+
+    /**
+     * The stub callables that will handle processes.
+     *
+     * @var array<int, callable(\Illuminate\Console\Process): (\Illuminate\Console\Contracts\ProcessResult|null)>
+     */
+    protected $stubs = [];
+
+    /**
+     * The callbacks that should execute before the process starts.
+     *
+     * @var array<int, callable(\Illuminate\Console\Process): mixed>
+     */
+    protected $beforeStartCallbacks = [];
+
+    /**
+     * Creates a new Pending Process instance.
+     *
+     * @param  \Illuminate\Console\Process\Factory  $factory
+     * @return void
+     */
+    public function __construct($factory)
+    {
+        $this->factory = $factory;
+    }
+
+    /**
+     * Ensures the process's run is delayed.
+     *
+     * @return $this
+     */
+    public function delayStart()
+    {
+        $this->delayStart = true;
+
+        return $this;
+    }
+
+    /**
+     * Dump the process and end the script before start the process.
+     *
+     * @return $this
+     */
+    public function dd()
+    {
+        return $this->beforeStart(fn ($process) => dd($process));
+    }
+
+    /**
+     * Dump the process before start the process.
+     *
+     * @return $this
+     */
+    public function dump()
+    {
+        return $this->beforeStart(fn ($process) => dump($process));
+    }
+
+    /**
+     * Register a stub callable that will intercept processes and be able to return stub process result.
+     *
+     * @param  array<int, callable(\Illuminate\Console\Process): \Illuminate\Console\Contracts\ProcessResult>  $callbacks
+     * @return $this
+     */
+    public function stubs($callbacks)
+    {
+        $this->stubs = $callbacks;
+
+        return $this;
+    }
+
+    /**
+     * Sets the process's command.
+     *
+     * @param  array<array-key, string>|string  $command
+     * @return $this
+     */
+    public function command($command)
+    {
+        return tap($this, fn () => $this->command = $command);
+    }
+
+    /**
+     * Sets the process's path.
+     *
+     * @param  string  $path
+     * @return $this
+     */
+    public function path($path)
+    {
+        return tap($this, fn () => $this->path = $path);
+    }
+
+    /**
+     * Sets the process's timeout.
+     *
+     * @param  float|null  $timeout
+     * @return $this
+     */
+    public function timeout($timeout)
+    {
+        return tap($this, fn () => $this->timeout = $timeout);
```

Comment:
Now this is not a draft anymore, could you revisit it? I am copying my last comment below to avoid explaining it again. Thanks =)

---

I do use `tap(...)` a lot on my projects, and the usage in `run()` method is something I think `tap()` brings simplicity.

But here and in the `withArgument` method, I don't see the advantage of using it. No temporary variables would be needed, both are simple assignments, and I know discussing micro-optimization is out of question, but it adds two function calls for no benefit (in my opinion).

Also I think in such cases it is actually more complex to skim through, and I am not aware of any other *"wither"* or fluent setter in the framework code which use this code style.

Reply:
Thanks for your comment. We are going to keep the `tap` for now.